### PR TITLE
Upload Coverage to Artifactory

### PR DESCRIPTION
**Pull request by mpdeimos**
Addresses issue [TS-23319](https://jira.cqse.eu/browse/TS-23319)

- [ ] Changes are tested adequately
- [ ] Agent's README.md updated in case of user-visible changes
- [ ] CHANGELOG.md updated
- [ ] [TGA Tutorial](https://docs.teamscale.com/tutorial/setting-up-tga-java) updated
- [ ] [TIA Tutorial](https://docs.teamscale.com/tutorial/tia-java/) updated

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

### CODE REVIEW

File: agent/src/main/java/com/teamscale/jacoco/agent/AgentBase.java [hunk 1]
```diff
@@ -87,9 +88,11 @@ public static void premain(String options, Instrumentation instrumentation) thro
 		HttpUtils.setShouldValidateSsl(agentOptions.shouldValidateSsl());
 
 		logger.info("Starting JaCoCo's agent");
-		org.jacoco.agent.rt.internal_43f5073.PreMain.premain(agentOptions.createJacocoAgentOptions(), instrumentation);
+		JacocoAgentBuilder agentBuilder = new JacocoAgentBuilder(agentOptions);
```
Comment (mpdeimos) [inline agent/src/main/java/com/teamscale/jacoco/agent/AgentBase.java:91]: For reviewer: created the AgentBuilder class to resolve AgentOptions growing over 300SLOC

File: agent/src/main/java/com/teamscale/jacoco/agent/git_properties/GitPropertiesLocatingTransformer.java [hunk 2]
```diff
@@ -46,9 +46,10 @@ public GitPropertiesLocatingTransformer(GitPropertiesLocator locator,
 
 		try {
 			CodeSource codeSource = protectionDomain.getCodeSource();
-			if (codeSource == null) {
+			if (codeSource == null || codeSource.getLocation() == null) {
```
Comment (mpdeimos) [inline agent/src/main/java/com/teamscale/jacoco/agent/git_properties/GitPropertiesLocatingTransformer.java:49]: Looks like a bug when executed on Java14

File: agent/src/main/java/com/teamscale/jacoco/agent/upload/HttpZipUploaderBase.java [hunk 3]
```diff
@@ -32,20 +33,33 @@
...
 
-		Retrofit retrofit = HttpUtils.createRetrofit(retrofitBuilder -> retrofitBuilder.baseUrl(uploadUrl));
-		api = getApi(retrofit);
+	/** Template method to configure the OkHttp Client. */
+	protected void configureOkHttp(OkHttpClient.Builder builder) {
 	}
 
 	/** Returns the API for creating request to the http uploader */
-	protected abstract T getApi(Retrofit retrofit);
+	protected T getApi() {
```
Comment (mpdeimos) [inline agent/src/main/java/com/teamscale/jacoco/agent/upload/HttpZipUploaderBase.java:54]: Needed to change to lazy-evaluation in order to allow better customization (::configureOkHttp)

File: agent/src/main/java/com/teamscale/jacoco/agent/upload/delay/DelayedUploader.java [hunk 4]
```diff
@@ -83,20 +82,24 @@ public String describe() {
...
 		} else {
 			logger.error("Tried to set upload commit multiple times (old uploader: {}, new commit: {})." +
-					" This is a programming error. Please report a bug.", wrappedUploader.describe(), commit);
+					" This is a programming error. Please report a bug.", wrappedUploader.describe(), information);
 		}
 	}
 
 	private void uploadCachedXmls() {
 		try {
+			if (!Files.isDirectory(cacheDir)) {
```
Comment (mpdeimos) [inline agent/src/main/java/com/teamscale/jacoco/agent/upload/delay/DelayedUploader.java:99]: For me the cacheDir was never set when the callback fired, so maybe looks like another Java14 issue

File: agent/src/main/java/com/teamscale/jacoco/agent/upload/artifactory/ArtifactoryConfig.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+		/** Constructor. */
+		public CommitInfo(String revision, CommitDescriptor commit) {
+			this.revision = revision;
+			this.commit = commit;
+		}
+
+		@Override
+		public String toString() {
+			return commit + "/" + revision;
```
Comment (DreierF) [inline agent/src/main/java/com/teamscale/jacoco/agent/upload/artifactory/ArtifactoryConfig.java:149]: Should we add a comment here where this is used?

Can we include "Commit "+commit+"/Revision "+revision or is this the path in artifactory?

File: agent/src/main/java/com/teamscale/jacoco/agent/upload/artifactory/ArtifactoryUploader.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	public ArtifactoryUploader(ArtifactoryConfig config, List<Path> additionalMetaDataFiles) {
+		super(config.url, additionalMetaDataFiles, IArtifactoryUploadApi.class);
+		this.artifactoryConfig = config;
+	}
+
+	@Override
+	protected void configureOkHttp(OkHttpClient.Builder builder) {
+		super.configureOkHttp(builder);
+		builder
+				.addInterceptor(HttpUtils.getBasicAuthInterceptor(artifactoryConfig.user, artifactoryConfig.password));
```
Comment (DreierF) [inline agent/src/main/java/com/teamscale/jacoco/agent/upload/artifactory/ArtifactoryUploader.java:33]: Formatting looks strange here

File: agent/README.md [hunk 7]
```diff
@@ -153,6 +153,18 @@ echo `git rev-parse --abbrev-ref HEAD`:`git --no-pager log -n1 --format="%ct000"
...
+ - `artifactory-user` (required for artifactory): The name of an artifactory user with write access.
+ - `artifactory-password` (required for artifactory): The password of the user.
+ - `artifactory-zip-path` (optional): The path within the stored ZIP file where the reports are stored.
+    Default is to store at root level.
+    This can be used to encode e.g. a partition name that is parsed later on via Teamscale Artifactory connector options. 
+ - `artifactory-git-properties-jar` (optional): Specify a Jar to search a `git.properties` file within.
+    If not specified, Git commit information is extracted from the first found `git.properties` file.
+    See `teamscale-git-properties-jar` for details. 
+ - `artifactory-git-properties-commit-date-format` (optional):
+    The Java data pattern `git.commit.time` is encoded with in `git.properties`. Defaults to `yyyy-MM-dd'T'HH:mm:ssZ`.
```
Comment (DreierF) [inline agent/README.md:167]: Do we really need the flexibility to specify the pattern?

File: agent/README.md [hunk 8]
```diff
@@ -153,6 +153,18 @@ echo `git rev-parse --abbrev-ref HEAD`:`git --no-pager log -n1 --format="%ct000"
...
+ - `artifactory-url`: the HTTP(S) url of the artifactory server to upload the reports to.
+    The URL may include a subpath on the artifactory server, e.g. `https://artifactory.acme.com/my-repo/my/subpath`.
+ - `artifactory-user` (required for artifactory): The name of an artifactory user with write access.
+ - `artifactory-password` (required for artifactory): The password of the user.
+ - `artifactory-zip-path` (optional): The path within the stored ZIP file where the reports are stored.
+    Default is to store at root level.
+    This can be used to encode e.g. a partition name that is parsed later on via Teamscale Artifactory connector options. 
+ - `artifactory-git-properties-jar` (optional): Specify a Jar to search a `git.properties` file within.
+    If not specified, Git commit information is extracted from the first found `git.properties` file.
+    See `teamscale-git-properties-jar` for details.
```
Comment (DreierF) [inline agent/README.md:165]: Do we need a second parameter for this? It basically does the same thing as the teamscale-git-properties-jar, right? What about renaming `teamscale-git-properties-jar` to be more generic?

File: agent/README.md (hunk 7 above)
Comment (mpdeimos) [reply agent/README.md:167]: Yes. At least optionally, as you can also specify in the git properties plugin, see here: https://github.com/git-commit-id/git-commit-id-maven-plugin/blob/master/maven/docs/using-the-plugin.md

Comment (mpdeimos) [reply agent/README.md:167]: and it has changed there a few releases back...

**Pull request merged**